### PR TITLE
Refactor DateAdapter component to CDK.

### PR DESCRIPTION
**Pull request by victoriaaa234**
Moved DateAdapter from material to CDK. 

DEPRECATIONS:
- Deprecated month constants from `src/lib/core`. Create month constants in your app instead, these were intended to be part of the non-exported testing module.
- `DateAdapter`, `NativeDateAdapter`, and `NativeDateModule` should now be imported from `@angular/cdk/datetime` rather than `@angular/material`
- `MAT_DATE_LOCALE` and `MAT_DATE_LOCAL_FACTORY` should be imported from `@angular/cdk/datetime` rather than `@angular/material`. The `MAT_` prefix has also been changed to `CDK_`

BREAKING CHANGES:
- Removed `MAT_DATE_LOCALE_PROVIDER` from `DateAdapter` component.

### CODE REVIEW

File: src/cdk/dateadapter/date-adapter.ts [hunk 1]
```diff
@@ -10,23 +10,23 @@ import {inject, InjectionToken, LOCALE_ID} from '@angular/core';
 import {Observable, Subject} from 'rxjs';
 
 /** InjectionToken for datepicker that can be used to override default locale code. */
-export const MAT_DATE_LOCALE = new InjectionToken<string>('MAT_DATE_LOCALE', {
+export const DATE_LOCALE = new InjectionToken<string>('DATE_LOCALE', {
```
Comment (jelbourn) [inline src/cdk/dateadapter/date-adapter.ts]: I'm worried that this is easily confused with Angular's own locale token. It might be worthwhile to make this "CDK_DATE_LOCALE" so that it's at least unambiguous.

File: src/material-moment-adapter/adapter/moment-date-adapter.spec.ts [hunk 2]
```diff
@@ -8,7 +8,8 @@
 
 import {LOCALE_ID} from '@angular/core';
 import {async, inject, TestBed} from '@angular/core/testing';
-import {DateAdapter, DEC, FEB, JAN, MAR, MAT_DATE_LOCALE} from '@angular/material/core';
+import {DateAdapter, DATE_LOCALE} from '@angular/cdk/dateadapter';
+import { DEC, FEB, JAN, MAR } from '@angular/cdk/testing';
```
Comment (jelbourn) [inline src/material-moment-adapter/adapter/moment-date-adapter.spec.ts]: Omit spaces in braces

Comment (rafaelss95) [reply src/material-moment-adapter/adapter/moment-date-adapter.spec.ts]: @jelbourn Does it worth to have a tslint rule that enforces consistent spaces? Or maybe better, have you ever considered to use [prettier](https://github.com/prettier/prettier)?

Comment (jelbourn) [reply src/material-moment-adapter/adapter/moment-date-adapter.spec.ts]: All of Google (and Angular core) actually uses clang-format for JS/TS internally, so if we were to do anything it would probably be that. At the beginning of this project we decided not to use it because it wrecks havoc on unit test indenting, and I'm not sure that's really any better than it was in 2016.

File: src/cdk/datetime/dateadapter/date-adapter.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+CDK DateAdapter component gives a foundation for more concrete material DateAdapter implementations.
+
+### Usages of CDK DateAdapter
+The base CDK version of the DateAdapter primarily manages the date functionality of the datepicker. In the CDK, there
```
Comment (tinayuangao) [inline src/cdk/datetime/dateadapter/date-adapter.md]: nit: add backtick for `DateAdapter`?

File: src/cdk/datetime/dateadapter/date-adapter.ts [hunk 4]
```diff
@@ -10,23 +10,23 @@ import {inject, InjectionToken, LOCALE_ID} from '@angular/core';
 import {Observable, Subject} from 'rxjs';
 
 /** InjectionToken for datepicker that can be used to override default locale code. */
-export const MAT_DATE_LOCALE = new InjectionToken<string>('MAT_DATE_LOCALE', {
```
Comment (tinayuangao) [inline src/cdk/datetime/dateadapter/date-adapter.ts]: How about the users who are using the `MAT_DATE_LOCALE`?

File: src/cdk/datetime/dateadapter/date-adapter.ts [hunk 5]
```diff
@@ -10,23 +10,23 @@ import {inject, InjectionToken, LOCALE_ID} from '@angular/core';
...
 
 /**
- * No longer needed since MAT_DATE_LOCALE has been changed to a scoped injectable.
+ * No longer needed since CDK_DATE_LOCALE has been changed to a scoped injectable.
  * If you are importing and providing this in your code you can simply remove it.
  * @deprecated
  * @deletion-target 7.0.0
  */
-export const MAT_DATE_LOCALE_PROVIDER = {provide: MAT_DATE_LOCALE, useExisting: LOCALE_ID};
+export const CDK_DATE_LOCALE_PROVIDER = {provide: CDK_DATE_LOCALE, useExisting: LOCALE_ID};
```
Comment (tinayuangao) [inline src/cdk/datetime/dateadapter/date-adapter.ts]: If this is deprecated, we don't have to move it into CDK

File: src/cdk/datetime/dateadapter/date-adapter.ts (hunk 4 above)
Comment (victoriaaa234) [reply src/cdk/datetime/dateadapter/date-adapter.ts]: I changed all of the dependencies on `MAT_DATE_LOCALE` to `CDK_DATE_LOCALE`

Comment (tinayuangao) [reply src/cdk/datetime/dateadapter/date-adapter.ts]: But `MAT_DATE_LOCALE` is a public API and may be used by external users.

Comment (victoriaaa234) [reply src/cdk/datetime/dateadapter/date-adapter.ts]: Would it be better to leave this as `MAT_DATE_LOCALE` in the DateAdapter CDK then?

File: src/material-moment-adapter/adapter/moment-date-adapter.spec.ts (hunk 2 above)
Comment (rafaelss95) [reply src/material-moment-adapter/adapter/moment-date-adapter.spec.ts]: And what about a rule? It can surely reduce the noise on PRs over whitespaces.

Comment (jelbourn) [reply src/material-moment-adapter/adapter/moment-date-adapter.spec.ts]: Probably, yeah. Feel free to send a PR if you're interested!

File: src/cdk/datetime/dateadapter/date-adapter.ts (hunk 4 above)
Comment (mmalerba) [reply src/cdk/datetime/dateadapter/date-adapter.ts]: In order to make this backward compatible, we have to re-export these from `src/lib/core/datetime/public-api.ts`:

```ts
export {
  DateAdapter, 
  CDK_DATE_LOCALE as MAT_DATE_LOCALE, 
  CDK_DATE_LOCALE_FACTORY as MAT_DATE_LOCALE_FACTORY
} from '@angular/cdk/datetime'
```

File: src/cdk/datetime/BUILD.bazel [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  bootstrap = ["//:web_test_bootstrap_scripts"],
+  # Do not sort
+  deps = [
+    "//:tslib_bundle",
+    "//:angular_bundles",
+    "//:angular_test_bundles",
+    "//test:angular_test_init",
+    ":datetime_test_sources",
+  ],
+)
```
Comment (mmalerba) [inline src/cdk/datetime/BUILD.bazel]: nit: add newline at end

File: src/cdk/datetime/dateadapter/date-adapter.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
+CDK `DateAdapter` component gives a foundation for more concrete material `DateAdapter` implementations.
```
Comment (mmalerba) [inline src/cdk/datetime/dateadapter/date-adapter.md]: the `dateadapter` directory isn't necessary, you can just put these files directly under `datetime`.

File: src/cdk/datetime/dateadapter/date-adapter.ts [hunk 8]
```diff
@@ -10,24 +10,16 @@ import {inject, InjectionToken, LOCALE_ID} from '@angular/core';
...
   return inject(LOCALE_ID);
 }
 
-/**
- * No longer needed since MAT_DATE_LOCALE has been changed to a scoped injectable.
- * If you are importing and providing this in your code you can simply remove it.
- * @deprecated
- * @deletion-target 7.0.0
- */
-export const MAT_DATE_LOCALE_PROVIDER = {provide: MAT_DATE_LOCALE, useExisting: LOCALE_ID};
```
Comment (mmalerba) [inline src/cdk/datetime/dateadapter/date-adapter.ts]: technically we can't remove this until the 7.0 release, though this PR will not go out sooner than that, so this is probably fine

File: src/cdk/datetime/datetime-module.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import {NgModule} from '@angular/core';
+import {PlatformModule} from '@angular/cdk/platform';
+
+@NgModule({
+    imports: [PlatformModule],
+    providers: [
+        {provide: DateAdapter, useClass: NativeDateAdapter},
+    ],
+})
+export class NativeDateModule {}
```
Comment (mmalerba) [inline src/cdk/datetime/datetime-module.ts]: nit: newline

File: src/lib/core/datetime/public-api.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export {
+    DateAdapter,
+    NativeDateAdapter,
+    NativeDateModule,
+    CDK_DATE_LOCALE as MAT_DATE_LOCALE,
+    CDK_DATE_LOCALE_FACTORY as MAT_DATE_LOCAL_FACTORY,
+} from '@angular/cdk/datetime'
```
Comment (mmalerba) [inline src/lib/core/datetime/public-api.ts]: nit: newline

File: src/lib/core/public-api.ts [hunk 11]
```diff
@@ -18,5 +18,3 @@ export * from './label/label-options';
 export * from './ripple/index';
 export * from './selection/index';
 
-// TODO: don't have this
-export * from './testing/month-constants';
```
Comment (mmalerba) [inline src/lib/core/public-api.ts]: People might be importing from here, you need to re-export it:

```ts
export {JAN, FEB, ...} from '@angular/cdk/testing';
```

(only re-export the ones that are currently exported, not all of `@angular/cdk/testing`, and leave the comment to eventually remove)

File: src/lib/schematics/update/material/data/method-call-checks.json [hunk 12]
```diff
@@ -1,21 +1,4 @@
 [
-  {
```
Comment (mmalerba) [inline src/lib/schematics/update/material/data/method-call-checks.json]: Why did you make this change? was it somehow required for things to compile?

File: src/material-moment-adapter/adapter/index.ts [hunk 13]
```diff
@@ -7,11 +7,7 @@
  */
 
 import {NgModule} from '@angular/core';
-import {
-  DateAdapter,
-  MAT_DATE_LOCALE,
-  MAT_DATE_FORMATS
-} from '@angular/material';
+import {DateAdapter, MAT_DATE_LOCALE, MAT_DATE_FORMATS} from '@angular/material';
```
Comment (mmalerba) [inline src/material-moment-adapter/adapter/index.ts]:
```ts
import {DateAdapter} from '@angular/cdk/datetime';
import {MAT_DATE_LOCALE, MAT_DATE_FORMATS} from '@angular/material';
```

File: src/material-moment-adapter/adapter/moment-date-adapter.spec.ts [hunk 14]
```diff
@@ -8,7 +8,8 @@
 
 import {LOCALE_ID} from '@angular/core';
 import {async, inject, TestBed} from '@angular/core/testing';
-import {DateAdapter, DEC, FEB, JAN, MAR, MAT_DATE_LOCALE} from '@angular/material/core';
+import {DateAdapter, MAT_DATE_LOCALE} from '@angular/material';
```
Comment (mmalerba) [inline src/material-moment-adapter/adapter/moment-date-adapter.spec.ts]: Bring in `DateAdapter` from CDK

File: src/lib/schematics/update/material/data/method-call-checks.json (hunk 12 above)
Comment (victoriaaa234) [reply src/lib/schematics/update/material/data/method-call-checks.json]: I thought it wasn't necessary since the `NativeDateAdapter` exists in `src/cdk/datetime` now. I will put it back.

File: src/cdk/datetime/dateadapter/date-adapter.ts (hunk 8 above)
Comment (mmalerba) [reply src/cdk/datetime/dateadapter/date-adapter.ts]: Probably around November, we aim for ~6 months between major versions

Comment (mmalerba) [reply src/cdk/datetime/dateadapter/date-adapter.ts]: Do document this as a BREAKING CHANGE in the description though

File: src/lib/schematics/update/material/data/method-call-checks.json (hunk 12 above)
Comment (mmalerba) [reply src/lib/schematics/update/material/data/method-call-checks.json]: This code is used to automatically migrate people from 5 to 6. We'll eventually add similar transforms to try to automatically update peoples code to fix breaking changes from the datepicker refactor, but for now I think we should just leave this alone

File: src/lib/datepicker/datepicker.spec.ts [hunk 15]
```diff
@@ -6,20 +6,16 @@ import {
...
-  MAT_DATE_LOCALE,
-  MatNativeDateModule,
-  NativeDateModule,
   SEP,
-} from '@angular/material/core';
+} from '@angular/cdk/testing';
+import {Component, FactoryProvider, Type, ValueProvider, ViewChild} from '@angular/core';
+import {ComponentFixture, fakeAsync, flush, inject, TestBed} from '@angular/core/testing';
+import {FormControl, FormsModule, ReactiveFormsModule} from '@angular/forms';
+import {MatNativeDateModule, MAT_DATE_LOCALE, NativeDateModule} from '@angular/material/core';
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker.spec.ts]: import `NativeDateModule` from cdk

File: src/lib/core/public-api.ts (hunk 11 above)
Comment (mmalerba) [reply src/lib/core/public-api.ts]: It looks like you didn't make this change?

File: src/lib/datepicker/datepicker.spec.ts (hunk 15 above)
Comment (victoriaaa234) [reply src/lib/datepicker/datepicker.spec.ts]: Forgot about that one!

File: src/lib/core/public-api.ts (hunk 11 above)
Comment (victoriaaa234) [reply src/lib/core/public-api.ts]: We talked about how testing is not exposed externally and our solution was to put `BREAKING CHANGES:` in the description

Comment (victoriaaa234) [reply src/lib/core/public-api.ts]: We talked about removing this, but clarified that we should still have this anyway

File: src/lib/core/datetime/public-api.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export {
```
Comment (mmalerba) [inline src/lib/core/datetime/public-api.ts]: you can mark all of this as deprecated, people should start importing it from cdk instead

File: src/lib/core/datetime/index.ts [hunk 17]
```diff
@@ -6,26 +6,14 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {PlatformModule} from '@angular/cdk/platform';
 import {NgModule} from '@angular/core';
-import {DateAdapter} from './date-adapter';
 import {MAT_DATE_FORMATS} from './date-formats';
-import {NativeDateAdapter} from './native-date-adapter';
 import {MAT_NATIVE_DATE_FORMATS} from './native-date-formats';
+import {NativeDateModule} from '@angular/cdk/datetime';
```
Comment (mmalerba) [inline src/lib/core/datetime/index.ts]: Since you're already making significant changes here, we should update this to look like our other `index.ts` files. Basically this should only re-export `public-api.ts`, re-exports from elsewhere should be in `public-api.ts`, and the `MatNativeDateModule` definition should be in a file called `datetime-module.ts` (just like how you structured it for the cdk)

File: src/lib/core/public-api.ts [hunk 18]
```diff
@@ -18,5 +18,8 @@ export * from './label/label-options';
...
+/**
+ * @deprecated Use month constants in `src/cdk/testing` instead.
```
Comment (mmalerba) [inline src/lib/core/public-api.ts]: We actually don't expose these publicly, so people can't use them instead. They should just create their own month constants for their application

File: src/material-moment-adapter/BUILD.bazel [hunk 19]
```diff
@@ -8,7 +8,7 @@ ts_library(
   name = "material-moment-adapter",
   srcs = glob(["**/*.ts"], exclude=["**/*.spec.ts"]),
   module_name = "@angular/material-moment-adapter",
-  deps = ["//src/lib:material"],
+  deps = ["//src/lib:material", "//src/cdk/datetime"],
```
Comment (mmalerba) [inline src/material-moment-adapter/BUILD.bazel]:
```
deps = [
    "...",
    "...",
]
```

Comment (victoriaaa234) [reply src/material-moment-adapter/BUILD.bazel]: Completed nits!

File: .github/CODEOWNERS [hunk 20]
```diff
@@ -64,6 +64,8 @@
 /src/cdk/bidi/**                                   @jelbourn
 /src/cdk/coercion/**                               @jelbourn
 /src/cdk/collections/**                            @jelbourn @crisbeto @andrewseguin
+/src/cdk/datepicker/**                             @mmalerba
+/src/cdk/datetime/**                               @mmalerba
```
[the remark this answers is not in the document]
Comment (mmalerba) [reply .github/CODEOWNERS]: They were separate when it was all under material, but I agree, it makes sense to combine them.

[the remark this answers is not in the document]
Comment (victoriaaa234) [reply .github/CODEOWNERS]: Combined

File: src/cdk/datetime/datetime-module.ts [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+ */
+
+import {DateAdapter} from './date-adapter';
+import {NativeDateAdapter} from './native-date-adapter';
+import {NgModule} from '@angular/core';
+import {PlatformModule} from '@angular/cdk/platform';
+
+@NgModule({
+    imports: [PlatformModule],
+    providers: [
```
Comment (rafaelss95) [inline src/cdk/datetime/datetime-module.ts]: Too much indent..

File: src/cdk/datetime/tsconfig-build.json [hunk 22]
```diff
@@ -2,13 +2,13 @@
   "extends": "../tsconfig-build",
   "files": [
     "public-api.ts",
-    "../typings.d.ts"
+    "../typings.d.ts",
```
Comment (rafaelss95) [inline src/cdk/datetime/tsconfig-build.json]: Is this change intentional?

File: src/lib/core/datetime/datetime-module.ts [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+ */
+
+import {NgModule} from '@angular/core';
+import {MAT_DATE_FORMATS} from './date-formats';
+import {MAT_NATIVE_DATE_FORMATS} from './native-date-formats';
+import {NativeDateModule} from '@angular/cdk/datetime';
+
+@NgModule({
+    imports: [NativeDateModule],
+    providers: [{provide: MAT_DATE_FORMATS, useValue: MAT_NATIVE_DATE_FORMATS}],
```
Comment (rafaelss95) [inline src/lib/core/datetime/datetime-module.ts]: Too much indent..

File: src/lib/core/datetime/public-api.ts (hunk 10 above)
Comment (rafaelss95) [reply src/lib/core/datetime/public-api.ts]: Too much indent..

File: src/cdk/datetime/date-adapter.md [hunk 24]
```diff
@@ -0,0 +1,15 @@
+The CDK `DateAdapter` facilitates use of any date data structure for date and time-based components, such as the
+material datepicker, and gives a foundation for more concrete material `DateAdapter` implementations.
+
+### Usages of CDK `DateAdapter`
+The base CDK version of the `DateAdapter` primarily manages the date functionality of the datepicker. In the CDK, there
+already exists an extended version of the `DateAdapter`: `NativeDateAdapter`. Currently in material, there's also
+another way to use the `DateAdapter`: `MomentDateAdapter`. The user also has the choice of creating their own
```
Comment (rafaelss95) [inline src/cdk/datetime/date-adapter.md]: "The user... their..."

File: src/cdk/datetime/tsconfig-build.json (hunk 22 above)
Comment (victoriaaa234) [reply src/cdk/datetime/tsconfig-build.json]: I just wanted to add a comma for future additional files (if necessary)

File: src/lib/core/datetime/public-api.ts [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * @deprecated Import from `src/cdk/datetime` instead.
+ * @deletion-target 8.0.0
+ */
+export {
+    DateAdapter,
+    NativeDateAdapter,
+    NativeDateModule,
+    CDK_DATE_LOCALE as MAT_DATE_LOCALE,
+    CDK_DATE_LOCALE_FACTORY as MAT_DATE_LOCAL_FACTORY,
```
Comment (rafaelss95) [inline src/lib/core/datetime/public-api.ts]: Too much indent..

File: src/cdk/datetime/public-api.ts [hunk 26]
```diff
@@ -9,3 +9,5 @@
 export * from './date-adapter';
 export * from './datetime-module';
 export * from './native-date-adapter';
+export * from './calendar-view';
+export * from './datepicker-module';
```
Comment (amcdnl) [inline src/cdk/datetime/public-api.ts]: nit: should end w/ new line
